{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# running annotation protocol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import sys\n",
    "import csv\n",
    "import time\n",
    "import random\n",
    "import requests\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_dir = Path.cwd().parent\n",
    "data_dir = base_dir / 'data'\n",
    "\n",
    "reference_data = {\n",
    "    11676: {\"fasta\": data_dir / \"reference/K03455.1.fasta\",\n",
    "            \"gtf\": data_dir / \"reference/K03455.1.gtf\"},\n",
    "    11723: {\"fasta\": data_dir / \"reference/M33262.1.fasta\",\n",
    "            \"gtf\": data_dir / \"reference/M33262.1.gtf\"}\n",
    "}\n",
    "\n",
    "sequence_dir = data_dir / 'sequences'\n",
    "assert sequence_dir.exists(), f\"sequence_dir does not exist: {sequence_dir}\"\n",
    "\n",
    "outdir = data_dir / 'annotation'\n",
    "outdir.mkdir(exist_ok=True)\n",
    "\n",
    "accession_list_fname = data_dir / 'complete_sequences.2022.accessions'\n",
    "summary_file =data_dir / \"download.summary.tsv\"\n",
    "\n",
    "logfile = outdir / 'log.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# software links\n",
    "soft_dir = base_dir / 'soft'\n",
    "vira_bin = 'vira'\n",
    "snapper_bin = 'snapper'\n",
    "miniprot_bin = soft_dir / 'miniprot/miniprot'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "soft_dir = Path().resolve().parent / 'soft/vira'  # Note: removed extra 'vira'\n",
    "sys.path.append(str(soft_dir))\n",
    "\n",
    "from vira.classes.txgroup import Transcriptome, Gene, Bundle\n",
    "from vira.classes.transcript import Transcript, Object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_genome_score(query_gtf_fname, target_gtf_fname, junction_stats_fname):\n",
    "    # compute overall score for the quality of annotation\n",
    "\n",
    "    assert os.path.exists(query_gtf_fname), f\"query_gtf_fname does not exist: {query_gtf_fname}\"\n",
    "    assert os.path.exists(target_gtf_fname), f\"target_gtf_fname does not exist: {target_gtf_fname}\"\n",
    "    assert os.path.exists(junction_stats_fname), f\"junction_stats_fname does not exist: {junction_stats_fname}\"\n",
    "\n",
    "    # parse query_gtf_fname to get base stats (number of junctions, number of genes, number of transcripts, number of proteins)\n",
    "    qry_tome = Transcriptome()\n",
    "    qry_tome.build_from_file(query_gtf_fname)\n",
    "\n",
    "    trg_tome = Transcriptome()\n",
    "    trg_tome.build_from_file(target_gtf_fname)\n",
    "\n",
    "    qry_donors = set()\n",
    "    qry_acceptors = set()\n",
    "    for qry_tx in qry_tome:\n",
    "        for it in qry_tx.introns_it():\n",
    "            qry_donors.add(it[0]-1)\n",
    "            qry_acceptors.add(it[1])\n",
    "    \n",
    "    junctions_df = pd.read_csv(junction_stats_fname, sep='\\t')\n",
    "    matching_donors = 0\n",
    "    matching_acceptors = 0\n",
    "    for i, row in junctions_df.iterrows():\n",
    "        # compute score of the position based on consistency and similarity of the flanking sequence\n",
    "        consistency_score = row[\"map_consistency\"]\n",
    "        similarity_score = sum(a == b for a, b in zip(row[\"sequence\"], row[\"query_sequence\"])) / len(row[\"sequence\"])\n",
    "        total_pos_score = consistency_score * similarity_score\n",
    "        if row[\"type\"] == \"donor\" and row['query_position'] in qry_donors:\n",
    "            matching_donors += total_pos_score\n",
    "        if row[\"type\"] == \"acceptor\" and row['query_position'] in qry_acceptors:\n",
    "            matching_acceptors += total_pos_score\n",
    "\n",
    "    junction_score = (matching_donors + matching_acceptors) / (len(qry_donors) + len(qry_acceptors))\n",
    "\n",
    "    # next we need to process the gtf file to compute scores for transcripts and proteins\n",
    "    transcript_score = 0\n",
    "    miniprot_score = 0\n",
    "    guide_score = 0\n",
    "    for tx in trg_tome:\n",
    "        qry_tx = qry_tome.get_by_tid(tx.tid)\n",
    "        vira_tx_aln_score = tx.get_attr(\"vira_tx_aln_score\")\n",
    "        vira_tx_aln_score = float(vira_tx_aln_score) if vira_tx_aln_score is not None else 0\n",
    "\n",
    "        miniprot_identity = tx.get_attr(\"identity_miniprot\")\n",
    "        miniprot_identity = float(miniprot_identity) if miniprot_identity is not None else 0\n",
    "        miniprot_score += miniprot_identity\n",
    "\n",
    "        guide_identity = tx.get_attr(\"identity_guide\")\n",
    "        guide_identity = float(guide_identity) if guide_identity is not None else 0\n",
    "        guide_score += guide_identity\n",
    "\n",
    "        has_cds = 1 if qry_tx.has_cds() == tx.has_cds() else 0\n",
    "\n",
    "        transcript_score += (vira_tx_aln_score + miniprot_identity + guide_identity + has_cds) / 4\n",
    "\n",
    "    transcript_score = transcript_score / len(qry_tome)\n",
    "    miniprot_score = miniprot_score / len(qry_tome)\n",
    "    guide_score = guide_score / len(qry_tome)\n",
    "\n",
    "    genome_annotation_score = (junction_score + transcript_score) / 2\n",
    "\n",
    "    return {\n",
    "            \"genome_score\": genome_annotation_score,\n",
    "            \"junction_score\": junction_score,\n",
    "            \"transcript_score\": transcript_score,\n",
    "            \"miniprot_score\": miniprot_score,\n",
    "            \"guide_score\": guide_score\n",
    "            }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "K03455.1\n",
      "MH705157.1\n",
      "JQ403028.1\n",
      "DQ396400.1\n",
      "DQ676872.1\n",
      "MH746258.1\n",
      "MH746253.1\n",
      "KU168256.1\n",
      "MH705158.1\n",
      "MH705151.1\n",
      "MH705153.1\n",
      "MH705133.1\n",
      "KU168305.1\n",
      "MN153491.1\n",
      "KP718918.1\n",
      "KP718928.1\n",
      "MT349418.1\n",
      "FJ388893.1\n",
      "FJ388903.1\n",
      "FJ388909.1\n",
      "FJ388925.1\n",
      "FJ388938.1\n",
      "FJ388942.1\n",
      "FJ388943.1\n",
      "JF683759.1\n",
      "JF683760.1\n",
      "JF683767.1\n",
      "JF683779.1\n",
      "JF683782.1\n",
      "JF683783.1\n",
      "FJ670519.1\n",
      "FJ670523.1\n",
      "KY496622.1\n",
      "MF109677.1\n",
      "MF109427.1\n",
      "MF109613.1\n",
      "MF109623.1\n",
      "KT152846.1\n",
      "KT152839.1\n",
      "KT152840.1\n",
      "KT152841.1\n",
      "KT152844.1\n",
      "AF457052.1\n",
      "AF457053.1\n",
      "AF457055.1\n",
      "AF457066.1\n",
      "AF457067.1\n",
      "AF457068.1\n",
      "AF457069.1\n",
      "AF457070.1\n",
      "AF457077.1\n",
      "AF457079.1\n",
      "AF457080.1\n",
      "AF457081.1\n",
      "AF457083.1\n",
      "AF457084.1\n",
      "AF457086.1\n",
      "AF457089.1\n",
      "EU110095.1\n",
      "EU110088.1\n",
      "EU110092.1\n",
      "EU110094.1\n",
      "KT022360.1\n",
      "KT022361.1\n",
      "KT022363.1\n",
      "KT022364.1\n",
      "KT022365.1\n",
      "KT022367.1\n",
      "KT022368.1\n",
      "KT022369.1\n",
      "KT022370.1\n",
      "KT022372.1\n",
      "KT022373.1\n",
      "KT022374.1\n",
      "KT022375.1\n",
      "KT022376.1\n",
      "KT022377.1\n",
      "KT022378.1\n",
      "KT022380.1\n",
      "KT022381.1\n",
      "KT022382.1\n",
      "KT022383.1\n",
      "FJ623487.1\n",
      "FJ623481.1\n",
      "FJ623475.1\n",
      "FJ623476.1\n",
      "FJ623480.1\n",
      "FJ623485.1\n",
      "FJ623483.1\n",
      "FJ623488.1\n",
      "FJ623478.1\n",
      "FJ623482.1\n",
      "FJ623477.1\n",
      "FJ623486.1\n",
      "FJ623479.1\n",
      "OM824814.2\n",
      "OM824614.2\n",
      "MW443147.1\n",
      "MW443157.1\n",
      "MN791333.1\n",
      "MN791710.1\n",
      "KF716474.1\n",
      "KF716475.1\n",
      "OM825002.2\n",
      "OM825064.2\n",
      "OM825085.2\n",
      "AF539405.1\n",
      "AF004885.1\n",
      "AY322185.1\n",
      "AY322190.1\n",
      "AY322193.1\n",
      "AF457063.1\n",
      "AF457065.1\n",
      "AF457075.1\n",
      "KU749411.1\n",
      "KX232594.1\n",
      "KX232595.1\n",
      "KX232597.1\n",
      "KX232600.1\n",
      "KX232607.1\n",
      "KX232609.1\n",
      "KX232610.1\n",
      "KX232611.1\n",
      "KX232613.1\n",
      "KX232614.1\n",
      "KX232618.1\n",
      "KX232620.1\n",
      "KX232624.1\n",
      "KX232626.1\n",
      "MT942764.1\n",
      "MT942708.1\n",
      "MT942776.1\n",
      "MT942744.1\n",
      "MT942721.1\n",
      "MT942795.1\n",
      "MT942807.1\n",
      "JX236677.1\n",
      "JX236678.1\n",
      "KU749424.1\n",
      "MT942842.1\n",
      "MT942826.1\n",
      "MT942889.1\n",
      "MT942875.1\n",
      "MT942915.1\n",
      "MT942931.1\n",
      "MT942947.1\n",
      "KF716472.1\n",
      "MZ642267.1\n",
      "MZ642264.1\n",
      "MZ642261.1\n",
      "MZ642260.1\n",
      "MZ642262.1\n",
      "MZ642266.1\n",
      "MZ642263.1\n",
      "MZ642269.1\n",
      "AB253421.1\n",
      "AB287376.1\n",
      "AB287379.1\n",
      "AY713406.1\n",
      "MF373167.1\n",
      "MF373168.1\n",
      "MF373181.1\n",
      "MF373182.1\n",
      "MF373124.1\n",
      "AF069670.1\n",
      "AF069671.1\n",
      "AF069669.1\n",
      "AF069673.1\n",
      "AF107771.1\n",
      "AY253305.1\n",
      "AY253314.1\n",
      "KX907352.1\n",
      "KX907348.1\n",
      "KX907336.1\n",
      "KX907343.1\n",
      "KX907372.1\n",
      "KX907383.1\n",
      "KX907389.1\n",
      "KX907341.1\n",
      "KX907347.1\n",
      "KX907364.1\n",
      "KX907401.1\n",
      "KX907412.1\n",
      "KX907346.1\n",
      "KX907378.1\n",
      "KX907414.1\n",
      "KX907423.1\n",
      "KX907431.1\n",
      "KY658694.1\n",
      "KY658695.1\n",
      "KY658696.1\n",
      "MN791853.1\n",
      "MN791873.1\n",
      "OM825644.2\n",
      "MN791964.1\n",
      "AF361872.1\n",
      "AF361873.1\n",
      "JX236669.1\n",
      "JX236671.1\n",
      "JX236676.1\n",
      "KF716478.1\n",
      "KP109490.1\n",
      "MN791279.1\n",
      "MN791169.1\n",
      "MN791209.1\n",
      "KF859745.1\n",
      "KF716486.1\n",
      "M62320.1\n",
      "OP039451.1\n",
      "OP039454.1\n",
      "OP039455.1\n",
      "OP039467.1\n",
      "OP039468.1\n",
      "OP039473.1\n",
      "OP039474.1\n",
      "OP039477.1\n",
      "OP039479.1\n",
      "OP039481.1\n",
      "OP039485.1\n",
      "OP039384.1\n",
      "OP039408.1\n",
      "AB253429.1\n",
      "AB098333.1\n",
      "AF484507.1\n",
      "AF484508.1\n",
      "AF484509.1\n",
      "AF484512.1\n",
      "AF484478.1\n",
      "AF484493.1\n",
      "AB098331.1\n",
      "AB485632.1\n",
      "KJ948658.1\n",
      "KT183312.1\n",
      "MH705163.1\n",
      "AF286238.1\n",
      "GU201516.1\n",
      "AF286237.1\n",
      "AY521629.1\n",
      "AY521631.1\n",
      "AY521630.1\n",
      "AM000053.1\n",
      "AM000054.1\n",
      "MN485977.1\n",
      "KT983615.1\n",
      "FJ388906.1\n",
      "FJ388950.1\n",
      "FJ388951.1\n",
      "JF683763.1\n",
      "JF683798.1\n",
      "ON989220.1\n",
      "ON989294.1\n",
      "ON989303.1\n",
      "ON989335.1\n",
      "MF109697.1\n",
      "DQ207944.1\n",
      "EU861977.1\n",
      "EF545108.1\n",
      "JQ292895.1\n",
      "JQ292896.1\n",
      "JQ292897.1\n",
      "JQ292898.1\n",
      "JQ292899.1\n",
      "JQ292900.1\n",
      "JQ292891.1\n",
      "KF716491.1\n",
      "KF716492.1\n",
      "KY238327.1\n",
      "FJ864679.1\n",
      "JQ292893.1\n",
      "JQ292894.1\n",
      "JX500696.1\n",
      "JX500695.1\n",
      "JX500694.1\n",
      "MH330347.1\n",
      "MH330348.1\n",
      "MH330349.1\n",
      "MH330350.1\n",
      "MH330351.1\n",
      "MH330352.1\n",
      "MH330353.1\n",
      "MH330354.1\n",
      "MH330355.1\n",
      "MH330356.1\n",
      "MH330358.1\n",
      "MH330359.1\n",
      "MH330361.1\n",
      "MH330363.1\n",
      "MH330364.1\n",
      "MH330367.1\n",
      "MZ427747.1\n",
      "MG902950.1\n",
      "MG902951.1\n",
      "MH330337.1\n",
      "MH330339.1\n",
      "MH330340.1\n",
      "MH330341.1\n",
      "MH330342.1\n",
      "MH330344.1\n",
      "MH330345.1\n",
      "MH330346.1\n",
      "MH330370.1\n",
      "MH330371.1\n",
      "MH330372.1\n",
      "MH330373.1\n",
      "MH330374.1\n",
      "MH330375.1\n",
      "MH330376.1\n",
      "MH330377.1\n",
      "MH330378.1\n",
      "MN703135.1\n",
      "MT222945.1\n",
      "MT222946.1\n",
      "MT222947.1\n",
      "MN703136.1\n",
      "MT222948.1\n",
      "MN703137.1\n",
      "MH330379.1\n",
      "MH330380.1\n",
      "MH330381.1\n",
      "MZ427727.1\n",
      "MT222949.1\n",
      "MZ427755.1\n",
      "MT318157.1\n",
      "MT334907.1\n",
      "MZ427717.1\n",
      "MT334908.1\n",
      "MT334909.1\n",
      "MT334910.1\n",
      "MT334911.1\n",
      "MT334912.1\n",
      "MT334913.1\n",
      "MZ427723.1\n",
      "MZ427722.1\n",
      "MN703138.1\n",
      "MN703139.1\n",
      "MT334915.1\n",
      "MT334916.1\n",
      "MT334919.1\n",
      "MT334920.1\n",
      "MT334921.1\n",
      "MT334924.1\n",
      "MT334925.1\n",
      "MT334926.1\n",
      "MT334927.1\n",
      "MT334928.1\n",
      "MT334929.1\n",
      "MT334930.1\n",
      "MT334931.1\n",
      "MT334932.1\n",
      "MT369950.1\n",
      "MT369951.1\n",
      "MT369952.1\n",
      "MT369953.1\n",
      "MT369955.1\n",
      "MT369956.1\n",
      "MT369957.1\n",
      "MT369958.1\n",
      "MT369960.1\n",
      "MT369961.1\n",
      "MT369963.1\n",
      "MT369964.1\n",
      "MT369966.1\n",
      "MT369967.1\n",
      "MT369970.1\n",
      "MT369972.1\n",
      "MT369973.1\n",
      "MT369975.1\n",
      "MT369977.1\n",
      "MT369978.1\n",
      "MT369980.1\n",
      "MT369982.1\n",
      "MT369983.1\n",
      "MT369984.1\n",
      "MT369987.1\n",
      "MT369989.1\n",
      "MT369990.1\n",
      "MT369993.1\n",
      "MT369994.1\n",
      "MT369995.1\n",
      "MT369997.1\n",
      "MT369998.1\n",
      "OP056071.1\n",
      "MT334918.1\n",
      "MT369968.1\n",
      "MT369999.1\n",
      "MT370000.1\n",
      "MT370001.1\n",
      "MT370002.1\n",
      "MT370003.1\n",
      "MT370005.1\n",
      "MT370006.1\n",
      "MT370007.1\n",
      "MT370008.1\n",
      "MZ427726.1\n",
      "MZ427756.1\n",
      "MZ427762.1\n",
      "MZ427766.1\n",
      "MZ427767.1\n",
      "MZ427720.1\n",
      "MZ327293.1\n",
      "MZ327294.1\n",
      "MZ427769.1\n",
      "MZ427770.1\n",
      "MZ427749.1\n",
      "MZ427746.1\n",
      "MZ427731.1\n",
      "MZ427734.1\n",
      "MZ427744.1\n",
      "MZ427743.1\n",
      "MZ427738.1\n",
      "MZ427733.1\n",
      "MZ427710.1\n",
      "MH330357.1\n",
      "MN736699.1\n",
      "AF413987.1\n",
      "DQ823357.1\n",
      "DQ823366.1\n",
      "KU749399.1\n",
      "KU749400.1\n",
      "KU749401.1\n",
      "KY658681.1\n",
      "KU749402.1\n",
      "KU749403.1\n",
      "KU749404.1\n",
      "KU749405.1\n",
      "KU749406.1\n",
      "KY658682.1\n",
      "KU749407.1\n",
      "FJ388928.1\n",
      "KX389622.1\n",
      "MH078558.1\n",
      "MW353966.1\n",
      "MW353967.1\n",
      "KY968401.1\n",
      "KY968402.1\n",
      "KY968403.1\n",
      "KY968404.1\n",
      "AY037269.1\n",
      "KY968394.1\n",
      "KY968395.1\n",
      "KY968407.1\n",
      "KY968408.1\n",
      "DQ383746.1\n",
      "KY968396.1\n",
      "DQ383748.1\n",
      "DQ383749.1\n",
      "KY968397.1\n",
      "DQ383751.1\n",
      "DQ383752.1\n",
      "KY968398.1\n",
      "KY968399.1\n",
      "KY658686.1\n",
      "KY658683.1\n",
      "MH078530.1\n",
      "MN703127.1\n"
     ]
    }
   ],
   "source": [
    "# read summary tsv line by line using csv module\n",
    "\n",
    "# Initialize log file\n",
    "if not os.path.exists(logfile):\n",
    "    with open(logfile, \"w\", newline=\"\") as logFP:\n",
    "        writer = csv.writer(logFP, quoting=csv.QUOTE_MINIMAL)\n",
    "        writer.writerow([\"accid\", \"organism\", \"status\", \"genome_score\", \"junction_score\", \"transcript_score\", \"miniprot_score\", \"guide_score\"])\n",
    "\n",
    "with open(summary_file, \"r\") as inFP, open(logfile, \"a\", newline=\"\") as logFP:\n",
    "    reader = csv.reader(inFP, delimiter=\"\\t\")\n",
    "    log_writer = csv.writer(logFP)\n",
    "    # skip header\n",
    "    next(reader)\n",
    "    for lno, (accid, is_ref_genome, organism, taxid, fasta, gff, is_valid, ref_start, ref_end) in enumerate(reader):\n",
    "        is_ref_genome = int(is_ref_genome)\n",
    "        is_valid = int(is_valid)\n",
    "        gff = int(gff)\n",
    "        fasta = int(fasta)\n",
    "        ref_start = int(ref_start)\n",
    "        ref_end = int(ref_end)\n",
    "        \n",
    "        reference_fasta_fname = reference_data[int(taxid)][\"fasta\"]\n",
    "        reference_gtf_fname = reference_data[int(taxid)][\"gtf\"]\n",
    "        \n",
    "        print(accid)\n",
    "\n",
    "        if is_ref_genome:\n",
    "            continue\n",
    "\n",
    "        fasta = bool(fasta)\n",
    "        gff = bool(gff)\n",
    "        is_valid = bool(is_valid)\n",
    "        if not fasta or not gff or not is_valid:\n",
    "            log_writer.writerow([accid, organism, \"skipped\", \"0\", \"0\", \"0\", \"0\", \"0\"])\n",
    "            logFP.flush()\n",
    "            continue\n",
    "\n",
    "        # check if output file exists and remove if it does\n",
    "        if (outdir / accid / f\"{accid}.vira.gtf\").exists():\n",
    "            os.remove(outdir / accid / f\"{accid}.vira.gtf\")\n",
    "\n",
    "        acc_outdir = outdir / accid\n",
    "        acc_outdir.mkdir(exist_ok=True)\n",
    "        output_gtf_fname = acc_outdir / f\"{accid}.vira.gtf\"\n",
    "        stdout_fname = acc_outdir / f\"{accid}.vira.stdout\"\n",
    "        stderr_fname = acc_outdir / f\"{accid}.vira.stderr\"\n",
    "        tmp_dir = acc_outdir / f\"{accid}.tmp\"\n",
    "        \n",
    "        target_fasta_fname = sequence_dir / f\"{accid}.fasta\"\n",
    "        # set guide to None if gff is False\n",
    "        guide_gtf_fname = sequence_dir / f\"{accid}.gtf\" if gff else None\n",
    "\n",
    "        cmd = [vira_bin,\n",
    "               \"--annotation\",str(reference_gtf_fname),\n",
    "               \"--genome\",str(reference_fasta_fname),\n",
    "               \"--target\",str(target_fasta_fname),\n",
    "               \"--output\",str(output_gtf_fname),\n",
    "               \"--miniprot\",str(miniprot_bin),\n",
    "               \"--snapper\",str(snapper_bin),\n",
    "               \"--tmp-dir\",str(tmp_dir)]\n",
    "               \n",
    "        if guide_gtf_fname is not None:\n",
    "            cmd.extend([\"--guide\", str(guide_gtf_fname)])\n",
    "\n",
    "        try:\n",
    "            with open(stdout_fname, \"w\") as stdoutFP, open(stderr_fname, \"w\") as stderrFP:\n",
    "                subprocess.run(cmd, stdout=stdoutFP, stderr=stderrFP, check=True)\n",
    "            target_gtf_fname = output_gtf_fname\n",
    "            junction_stats_fname = str(output_gtf_fname).rsplit(\".\",1)[0] + \".junction_stats.tsv\"\n",
    "            scores = compute_genome_score(reference_gtf_fname, output_gtf_fname, junction_stats_fname)\n",
    "            log_writer.writerow([accid, organism, \"complete\", scores[\"genome_score\"], scores[\"junction_score\"], scores[\"transcript_score\"], scores[\"miniprot_score\"], scores[\"guide_score\"]])\n",
    "        except subprocess.CalledProcessError:\n",
    "            # Read the error message from the stderr file\n",
    "            with open(stderr_fname, \"r\") as stderrFP:\n",
    "                error_message = stderrFP.read().strip()\n",
    "            # Log only the last line of the error message\n",
    "            log_writer.writerow([accid, organism, \"failed\", \"0\", \"0\", \"0\", \"0\", \"0\"])\n",
    "        finally:\n",
    "            logFP.flush()\n",
    "\n",
    "        if tmp_dir.exists():\n",
    "            for f in tmp_dir.iterdir():\n",
    "                f.unlink()\n",
    "            tmp_dir.rmdir()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hivAtlas",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
