{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# build a database for the web app populating it will all sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import sys\n",
    "import csv\n",
    "import time\n",
    "import shutil\n",
    "import random\n",
    "import sqlite3\n",
    "import requests\n",
    "import subprocess\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_dir = Path.cwd().parent\n",
    "data_dir = base_dir / 'data'\n",
    "\n",
    "reference_fasta_fname = data_dir / 'reference' / 'K03455.1.fasta'\n",
    "reference_gtf_fname = data_dir / 'reference' / 'K03455.1.gtf'\n",
    "\n",
    "sequence_dir = data_dir / 'sequences'\n",
    "assert sequence_dir.exists(), f\"sequence_dir does not exist: {sequence_dir}\"\n",
    "\n",
    "outdir = data_dir / 'annotation'\n",
    "outdir.mkdir(exist_ok=True)\n",
    "\n",
    "release_dir = Path.cwd().parent.parent / 'HIV_Atlas_Data'\n",
    "assert release_dir.exists(), f\"release_dir does not exist: {release_dir}\"\n",
    "release_data_dir = release_dir / 'data'\n",
    "release_data_dir.mkdir(exist_ok=True)\n",
    "release_db_file = release_dir / 'hiv_atlas.db.tsv'\n",
    "\n",
    "accession_list_fname = data_dir / 'complete_sequences.2022.accessions'\n",
    "summary_file =data_dir / \"download.summary.tsv\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_accid_from_fasta(fasta_fname):\n",
    "    with open(fasta_fname, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('>'):\n",
    "                return line.split()[0][1:]\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_genome_files(fasta_fname, gtf_fname, out_base_name):\n",
    "    \"\"\"Build the genome files for the database.\"\"\"\n",
    "\n",
    "    cmds = []\n",
    "\n",
    "    # create copies of fasta and gtf files\n",
    "    new_fasta_fname = f\"{out_base_name}.fasta\"\n",
    "    new_gtf_fname = f\"{out_base_name}.gtf\"\n",
    "    shutil.copy(fasta_fname, new_fasta_fname)\n",
    "    shutil.copy(gtf_fname, new_gtf_fname)\n",
    "\n",
    "    # create fasta index\n",
    "    cmds.append(f\"samtools faidx {new_fasta_fname}\")\n",
    "\n",
    "    # create sorted gff: gffread -F --keep-exon-attrs -o- K03455.1.gtf | sort -k1,1 -k4,4n > K03455.1.sorted.gff\n",
    "    gff_fname = outdir / f\"{out_base_name}.gff\"\n",
    "    cmds.append(f\"gffread -F --keep-exon-attrs -o- {new_gtf_fname} | sort -k1,1 -k4,4n > {gff_fname}\")\n",
    "\n",
    "    # compress the gff file with bgzip - keep the original\n",
    "    cmds.append(f\"bgzip -k {gff_fname}\")\n",
    "\n",
    "    # create tabix index\n",
    "    cmds.append(f\"tabix -p gff {gff_fname}.gz\")\n",
    "\n",
    "    # return dictionary of file names\n",
    "    return {\n",
    "        'fasta': new_fasta_fname,\n",
    "        'fai': f\"{new_fasta_fname}.fai\",\n",
    "        'gtf': new_gtf_fname,\n",
    "        'gff': f\"{gff_fname}.gz\",\n",
    "        'tbi': f\"{gff_fname}.gz.tbi\",\n",
    "        'cmds': \"\\n\".join(cmds)\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add K03455.1 as the reference genome\n",
    "genome_db_outdir = release_data_dir / 'K03455.1'\n",
    "genome_db_outdir.mkdir(exist_ok=True)\n",
    "\n",
    "genome_data = build_genome_files(reference_fasta_fname, reference_gtf_fname, genome_db_outdir / 'K03455.1')\n",
    "# write the commands to a file\n",
    "with open(release_data_dir / 'commands.sh', 'w') as outFP:\n",
    "    outFP.write(genome_data['cmds']+\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read summary tsv line by line using csv module\n",
    "with open(release_data_dir / 'commands.sh', 'a') as outFP:\n",
    "    with open(summary_file,\"r\") as inFP:\n",
    "        reader = csv.reader(inFP, delimiter=\"\\t\")\n",
    "        # skip header\n",
    "        next(reader)\n",
    "        for lno, (accid, fasta, gff) in enumerate(reader):\n",
    "            if accid == \"K03455\": # skip HXB2\n",
    "                continue\n",
    "\n",
    "            if (outdir / accid / f\"{accid}.vira.gtf\").exists():\n",
    "                acc_outdir = outdir / accid\n",
    "                acc_outdir.mkdir(exist_ok=True)\n",
    "                genome_fasta_fname = sequence_dir / f\"{accid}.fasta\"\n",
    "                genome_gtf_fname = acc_outdir / f\"{accid}.vira.gtf\"\n",
    "                \n",
    "                fasta_accid = get_accid_from_fasta(genome_fasta_fname)\n",
    "                assert fasta_accid is not None, f\"Could not get accession id from fasta file: {genome_fasta_fname}\"\n",
    "                \n",
    "                # create gff file, sort and compress and index\n",
    "                genome_db_outdir = release_data_dir / fasta_accid\n",
    "                genome_db_outdir.mkdir(exist_ok=True)\n",
    "\n",
    "                genome_data = build_genome_files(genome_fasta_fname, genome_gtf_fname, genome_db_outdir / fasta_accid)\n",
    "                # write the commands to a file\n",
    "                with open(release_data_dir / 'commands.sh', 'a') as outFP:\n",
    "                    outFP.write(genome_data['cmds']+\"\\n\")\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Existing database deleted.\n",
      "Database schema created.\n",
      "Inserted: K03455.1\n"
     ]
    }
   ],
   "source": [
    "# build\n",
    "\n",
    "def get_description(fasta_fname):\n",
    "    \"\"\"Get the description from the fasta file.\"\"\"\n",
    "    with open(fasta_fname) as inFP:\n",
    "        for line in inFP:\n",
    "            if line.startswith('>'):\n",
    "                return line.strip().split(\" \",1)[-1]\n",
    "            \n",
    "    return \"\"\n",
    "\n",
    "\n",
    "# initialize manual db file\n",
    "db_tsvFP = open(release_db_file, \"w\") # format: accid, is_reference, description, dir\n",
    " \n",
    "genome_db_outdir = release_data_dir / 'K03455.1'\n",
    "genome_db_outdir.mkdir(exist_ok=True)\n",
    "\n",
    "reference_genome = {\n",
    "    'accession_id': 'K03455.1',\n",
    "    'description': 'Human immunodeficiency virus type 1 (HXB2), complete genome; HIV1/HTLV-III/LAV reference genome',\n",
    "    'dir': str(genome_db_outdir),\n",
    "    'is_reference': True\n",
    "}\n",
    "db_tsvFP.write(f\"{reference_genome['accession_id']}\\t1\\t{reference_genome['description']}\\n\")\n",
    "\n",
    "# read summary tsv line by line using csv module\n",
    "with open(summary_file,\"r\") as inFP:\n",
    "    reader = csv.reader(inFP, delimiter=\"\\t\")\n",
    "    # skip header\n",
    "    next(reader)\n",
    "    for lno, (accid, fasta, gff) in enumerate(reader):\n",
    "        if accid == \"K03455\": # skip HXB2\n",
    "            continue\n",
    "\n",
    "        if (outdir / accid / f\"{accid}.vira.gtf\").exists():\n",
    "            acc_outdir = outdir / accid\n",
    "            acc_outdir.mkdir(exist_ok=True)\n",
    "            genome_fasta_fname = sequence_dir / f\"{accid}.fasta\"\n",
    "            genome_gtf_fname = acc_outdir / f\"{accid}.vira.gtf\"\n",
    "            \n",
    "            fasta_accid = get_accid_from_fasta(genome_fasta_fname)\n",
    "            assert fasta_accid is not None, f\"Could not get accession id from fasta file: {genome_fasta_fname}\"\n",
    "            \n",
    "            # create gff file, sort and compress and index\n",
    "            genome_db_outdir = release_data_dir / fasta_accid\n",
    "            # build \n",
    "            genome = {\n",
    "                'accession_id': fasta_accid,\n",
    "                'description': get_description(genome_fasta_fname),\n",
    "                'dir': str(genome_db_outdir),\n",
    "                'is_reference': False\n",
    "            }\n",
    "\n",
    "            db_tsvFP.write(f\"{genome['accession_id']}\\t0\\t{genome['description']}\\n\")\n",
    "\n",
    "db_tsvFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cleanup uncompressed gff and gtf files\n",
    "for accid in os.listdir(release_data_dir):\n",
    "    accdir = release_data_dir / accid\n",
    "    if not accdir.is_dir():\n",
    "        continue\n",
    "\n",
    "    for fname in os.listdir(accdir):\n",
    "        if fname.endswith(\".gff\") or fname.endswith(\".gtf\"):\n",
    "            os.remove(accdir / fname)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
