{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute conservation across each genome for every position on the reference genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import sys\n",
    "import csv\n",
    "import time\n",
    "import random\n",
    "import requests\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_dir = Path.cwd().parent\n",
    "data_dir = base_dir / 'data'\n",
    "\n",
    "reference_fasta_fname = data_dir / 'reference' / 'K03455.1.fasta'\n",
    "reference_gtf_fname = data_dir / 'reference' / 'K03455.1.gtf'\n",
    "\n",
    "sequence_dir = data_dir / 'sequences'\n",
    "assert sequence_dir.exists(), f\"sequence_dir does not exist: {sequence_dir}\"\n",
    "\n",
    "annotation_dir = data_dir / 'annotation'\n",
    "assert annotation_dir.exists(), f\"annotation_dir does not exist: {annotation_dir}\"\n",
    "\n",
    "outdir = data_dir / 'conservation'\n",
    "outdir.mkdir(exist_ok=True)\n",
    "\n",
    "accession_list_fname = data_dir / 'complete_sequences.2022.accessions'\n",
    "summary_file =data_dir / \"download.summary.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# software links\n",
    "soft_dir = base_dir / 'soft'\n",
    "mafft_bin = \"mafft\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load list of genomes\n",
    "\n",
    "# read summary tsv line by line using csv module\n",
    "genomes = {}\n",
    "with open(summary_file,\"r\") as inFP:\n",
    "    reader = csv.reader(inFP, delimiter=\"\\t\")\n",
    "    # skip header\n",
    "    next(reader)\n",
    "    for lno, (accid, fasta, gff) in enumerate(reader):\n",
    "        if accid == \"K03455\": # skip HXB2\n",
    "            continue\n",
    "\n",
    "        if (annotation_dir / accid / f\"{accid}.vira.gtf\").exists():\n",
    "            acc_outdir = annotation_dir / accid\n",
    "            acc_outdir.mkdir(exist_ok=True)\n",
    "            genome_fasta_fname = sequence_dir / f\"{accid}.fasta\"\n",
    "            genome_gtf_fname = acc_outdir / f\"{accid}.vira.gtf\"\n",
    "            \n",
    "            genomes[accid] = (genome_fasta_fname, genome_gtf_fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# build a file of all sequence for alignments\n",
    "qry_fasta_fname = outdir / 'all_sequences.fasta'\n",
    "with open(qry_fasta_fname, 'w') as outFP:\n",
    "    for accid, (genome_fasta_fname, genome_gtf_fname) in genomes.items():\n",
    "        with open(genome_fasta_fname, 'r') as inFP:\n",
    "            for line in inFP:\n",
    "                outFP.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --6merpair --addfragments /home/avaraby1/HIV_Atlas/data/conservation/all_sequences.fasta /home/avaraby1/HIV_Atlas/data/reference/K03455.1.fasta > /home/avaraby1/HIV_Atlas/data/conservation/all_sequences.msa\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 3090\n",
      "ppenalty_ex = -10\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 1\n",
      "njobc = 2\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "\n",
      "There are 39208 ambiguous characters\n",
      "    1 / 1\n",
      "done.\n",
      "\n",
      "fTEP 499 / 3090                    \n",
      "STEP 500 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 600 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 700 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 800 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 900 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 1000 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 1100 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 1200 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 1300 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 1400 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 1500 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 1600 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 1700 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 1800 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 1900 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 2000 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 2100 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 2200 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 2300 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 2400 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 2500 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 2600 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 2700 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 2800 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 2900 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "STEP 3000 / 3090                    \n",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                       \n",
      "\n",
      "addsingle (nuc) Version 7.526\n",
      "alg=A, model=DNA200 (2), 1.53 (4.59), -0.00 (-0.00), noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-NS-fragment (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# run mafft\n",
    "# mafft --6merpair --addfragments othersequences referencesequence > output\n",
    "msa_fname = outdir / 'all_sequences.msa'\n",
    "cmd = [mafft_bin, '--6merpair', '--addfragments', str(qry_fasta_fname), str(reference_fasta_fname)]\n",
    "print(\" \".join(cmd) + \" > \" + str(msa_fname))\n",
    "# with open(msa_fname, 'w') as f:\n",
    "#     subprocess.run(cmd, stdout=f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load a matrix from the msa where columns are positions on the reference genome\n",
    "# rows are genomes\n",
    "# and values are tuple of (genome position, nucleotide) at that position\n",
    "\n",
    "def load_msa_as_matrix(fasta_file):\n",
    "    # Load the alignment\n",
    "    sequences = list(SeqIO.parse(fasta_file, \"fasta\"))\n",
    "    alignment_length = len(sequences[0].seq)\n",
    "    for seq_record in sequences:\n",
    "        if len(seq_record.seq) != alignment_length:\n",
    "            raise ValueError(\"All sequences in the alignment must be the same length.\")\n",
    "\n",
    "    # Create a matrix where each row corresponds to a genome\n",
    "    matrix = []\n",
    "    idx = {} # map of accession to row in matrix\n",
    "    for seq_index, seq_record in enumerate(sequences):\n",
    "        idx[seq_record.id] = seq_index\n",
    "        row = []\n",
    "        genome_position = 0  # Tracks the position in the original genome\n",
    "        for align_position, nucleotide in enumerate(seq_record.seq):\n",
    "            if nucleotide != \"-\":  # Skipping gaps\n",
    "                genome_position += 1\n",
    "            row.append((genome_position, nucleotide))\n",
    "        matrix.append(row)\n",
    "\n",
    "    return matrix, idx\n",
    "\n",
    "msa_mat, msa_idx = load_msa_as_matrix(msa_fname)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now cleanup the matrix by removing all column which do not contain reference genome nucleotide\n",
    "\n",
    "def cleanup_matrix(mat, ref_idx):\n",
    "    assert len(mat) > 0, \"Matrix must have at least one row.\"\n",
    "    assert ref_idx > 0 or ref_idx < len(mat), \"Reference index is out of bounds.\"\n",
    "\n",
    "    reference_row = mat[ref_idx]\n",
    "    columns_to_keep = [i for i in range(len(reference_row)) if reference_row[i][1] != '-']\n",
    "\n",
    "    clean_mat = []\n",
    "    for row in mat:\n",
    "        clean_row = [row[i] for i in columns_to_keep]\n",
    "        clean_mat.append(clean_row)\n",
    "\n",
    "    return clean_mat\n",
    "\n",
    "\n",
    "ref_msa_idx = msa_idx[\"K03455.1\"]\n",
    "clean_msa_mat = cleanup_matrix(msa_mat, ref_msa_idx)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute simple conservation: for each position - count the number of genoems where the nucleotide is the same as the reference genome\n",
    "def compute_simple_conservation(mat, ref_idx):\n",
    "    assert len(mat) > 0, \"Matrix must have at least one row.\"\n",
    "    assert ref_idx > 0 or ref_idx < len(mat), \"Reference index is out of bounds.\"\n",
    "    ref_row = mat[ref_idx]\n",
    "    \n",
    "    conservation_counts = []\n",
    "    \n",
    "    # Iterate over each column (position)\n",
    "    for col_index in range(len(ref_row)):\n",
    "        ref_nuc = ref_row[col_index][1]\n",
    "        count = 0\n",
    "        \n",
    "        for row in mat:\n",
    "            if row[col_index][1] == ref_nuc:\n",
    "                count += 1\n",
    "                \n",
    "        conservation_counts.append(count)\n",
    "    \n",
    "    return conservation_counts\n",
    "\n",
    "simple_conservation = compute_simple_conservation(clean_msa_mat, ref_msa_idx)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write result as bed format\n",
    "bed_fname = outdir / 'K03455.1.conservation.simple.all.bed'\n",
    "with open(bed_fname, 'w') as outFP:\n",
    "    for i, count in enumerate(simple_conservation):\n",
    "        outFP.write(f\"K03455.1\\t{i}\\t{i+1}\\t-\\t{count}\\t+\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now we shall also consider the donor/acceptor information\n",
    "# specifically, for each donor and acceptor site on the reference genome\n",
    "# we should find out the number of genomes where the corresponding position also has a donor/acceptor site\n",
    "\n",
    "# load all genome annotations"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
